### PR TITLE
Sparse CSR: Add `torch.randn_like`

### PR DESCRIPTION
**Pull request by IvanYashchuk**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #68084
* __->__ #68083
* #71103

This PR adds support for `torch.randn_like(sparse_csr_tensor)`.
It creates a new sparse csr tensor with same indices but different values that are normally distributed.

In addition `.normal_()` and `torch.empty_like` were implemented because `randn_like` is a composite of these two functions.

cc @nikitaved @pearu @cpuhrsch @IvanYashchuk

Differential Revision: [D33511280](https://our.internmc.facebook.com/intern/diff/D33511280)

### CODE REVIEW

File: test/test_sparse_csr.py [hunk 1]
```diff
@@ -1070,6 +1074,23 @@ def test_coo_csr_conversion(self, device, dtype):
...
+
+        for sample in samples:
+            assert isinstance(sample.input, torch.Tensor)
+            if sample.input.ndim != 2:
+                continue
+
+            expected = op(sample.input)
+            assert torch.is_tensor(expected)
+            output = op(sample.input.to_sparse_csr())
+            assert torch.is_tensor(output)
```
Comment (IvanYashchuk) [inline test/test_sparse_csr.py:1092]: We can actually compare `expected` and `output.to_dense()` with `self.assertEqual`. It passes for `randn_like`, but of course fails for `empty_like`.

File: aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp [hunk 2]
```diff
@@ -59,6 +59,14 @@ void convert_indices_from_coo_to_csr_cpu(const Tensor& result, const Tensor& inp
     data_out[i] = static_cast<output_t>(numel);
 }
 
+template <typename F, typename ...Args>
+Tensor& unary_inplace_op_(F op_, Tensor& self, Args&&... args) {
+  TORCH_INTERNAL_ASSERT(self.is_sparse_csr());
+  auto values = self.values();
+  (self.values().*op_)(std::forward<Args>(args)...);
```
Comment (krshrimali) [inline aten/src/ATen/native/sparse/SparseCsrTensorMath.cpp:66]: Do we need `auto values = self.values()` here? I feel it's redundant?

File: aten/src/ATen/native/TensorFactories.cpp [hunk 3]
```diff
@@ -272,6 +272,18 @@ Tensor empty_like(
     return result;
   }
 
+  if (options.layout() == kSparseCsr && self.is_sparse_csr()) {
+    auto result = at::native::_sparse_csr_tensor_unsafe(
+        self.crow_indices().clone(),
+        self.col_indices().clone(),
+        at::empty(self.values().sizes(), options.layout(kStrided)),
+        self.sizes(),
+        self.scalar_type(),
```
Comment (krshrimali) [inline aten/src/ATen/native/TensorFactories.cpp:281]: Post offline discussion with @IvanYashchuk - This should be `dtype` if provided, else `self.scalar_type()`.

Comment (IvanYashchuk) [reply aten/src/ATen/native/TensorFactories.cpp:281]: This should be `options.dtype()`, the device should also be obtained from `options`.

**Comment by cpuhrsch** [comment]
On a high level note: In light of our recent decisions around sparsity as transparent layouts and constraining unspecified values to mean zero, how do we best fit operators like randn_like or normal into it?

File: test/test_sparse_csr.py (hunk 1 above)
Comment (pearu) [reply test/test_sparse_csr.py:1077]: > It passes for randn_like.

I think this happens by accident (if `sample.input` is a random dense tensor it likely does not contain zeros, that is, it is full tensor, then also `to_sparse_csr` results in a full tensor) and not in general (if `sample.input` would not be a full tensor then `assertEqual(output.to_dense(), expected)` will fail at unspecified element indices).

Comment (IvanYashchuk) [reply test/test_sparse_csr.py:1077]: It was a carefully designed test of creating a sparse representation of a surely dense tensor, so not by accident 😄

Comment (pearu) [reply test/test_sparse_csr.py:1077]: Not sure what do you mean by a carefully designed test considering that the domain of a number of operations that support sparse csr inputs, contains 0.

File: test/test_sparse_csr.py [hunk 4]
```diff
@@ -1070,6 +1074,23 @@ def test_coo_csr_conversion(self, device, dtype):
...
+    def test_sparse_csr_no_error(self, device, dtype, op):
+        samples = op.sample_inputs(device, dtype)
+
+        if len(samples) == 0:
+            self.skipTest("Skipped! No sample inputs!")
+
+        for sample in samples:
+            assert isinstance(sample.input, torch.Tensor)
+            if sample.input.ndim != 2:
+                continue
```
Comment (pearu) [inline test/test_sparse_csr.py:1072]: With https://github.com/pytorch/pytorch/blob/6c4437118b88c5756b0a192263fb8ed40cfe2800/torch/testing/_internal/common_methods_invocations.py#L1184-L1186, this if-block is redudant.

**Review by cpuhrsch — APPROVED** [review]

**Comment by cpuhrsch** [comment]
@cpuhrsch has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33511280).

**Pull request closed without merging**